### ISSUE TITLE
Automatically enable no-std if std isn't present

### ISSUE THREAD

**Issue by kayabaNerve** [rust-bitcoin/rust-bitcoin#2155]
Currently, `no-std` must be explicitly enabled when the de-facto approach to handling no-std/std is to check if std isn't enabled or if it is. Since features can't be *disabled* by Cargo once set, this forces all std consumers of a higher-level library which supports a no-std variant to *additionally* build the no-std code.

If no-std was assumed if std wasn't set, then std consumers wouldn't have to build no-std regardless.

This *may* enable removing the no-std feature, yet some people may want there to be both the no-std (currently core2-based) and std code present. That would still justify its existence.

Relevant #758, #2034, #2043, #2066.

**Comment by apoelstra** [comment]
We cannot do this, because `no-std` enables `core2`. If enabling `std` were to disable `core2` then `std` would no longer be an additive feature.

I definitely agree that having the `no-std` feature sucks. But it's not clear how to move forward on this.

**Comment by Kixunil** [comment]
This is effectively a duplicate of #758 and is being worked on. `hashbrown` was already removed, `core2` will be removed in #2066

**Comment by kayabaNerve** [comment]
@Kixunil While I agree 758 is similar and this could have been posted there, 758 never explicitly proposed removing no-std. It solely discusses its issues and proposes a new feature design (which may or may not still involve a no-std feature).  This was meant to hone in on the short-term goal of removing the no-std feature. Apologies if this wasn't optimally submitted.

Also, the cited PR maintains an independent `no-std` feature, so this isn't being directly worked on.

@apoelstra There's a distinction between noting core2 as a dependency, which would be required for all builds if the no-std feature is removed (a trade-off I'd prefer), and using core2 in bitcoin (and accordingly generating code involving it). With #2066, and the removal of core2, this discussion changes as then the sole purpose of the no-std feature is... to build two sets of code? To explicitly identify as no-std? I'm honestly unsure.

Regardless, I'd personally prefer the other side of the trade off regarding core2 now, and I see no reason why the above PR shouldn't incorporate this change. 

(And yes, I can move further discussion to #758 if preferred)

**Comment by apoelstra** [comment]
>which would be required for all builds if the no-std feature is removed (a trade-off I'd prefer)

I'm afraid this is a total non-starter. core2 is a huge dependency that none of us wants, and we were able to agree that it would be okay for people who explicitly opted into it because they're doing kinda-unstable things anyway. But even there, there's a push to remove it.

Unconditionally including `bitcoin_io` after #2066 is more interesting. The `bitcoin_io` crate is much smaller and maintained by us. I think I would prefer to have it be unconditional than to have this gross `no-std` feature. You're right that #2066 doesn't do this, but I think *after* #2066 we'd be able to accept a small/simple PR that does it. At least, I'd be in favor of it personally.

**Comment by Kixunil** [comment]
> 758 never explicitly proposed removing no-std

"Finer feature gating" implies breaking up `no-std` into sub-features, so logically `no-std` would be removed. But to be absolutely clear, I don't blame you for proposing it explicitly (you could've on that issue instead) and yes, that's what I intended.

> Also, the cited PR maintains an independent `no-std` feature, so this isn't being directly worked on.

It's probably the most important step towards removing it. As you can see it only depends on `*/alloc` which can then be renamed to `alloc` as it should be. Also note that gating everything on `alloc` is a lot of work, so it will take time. Feel free to help!

> the sole purpose of the no-std feature is...

To get MVP of `no-std` out ASAP - at the time it was written. People just didn't put in that much of work to split it up properly and feature gate individual APIs. And it's really a lot. Maybe 90% of the crate cannot be built without the IO traits for instance. (I tried.)

I agree with @apoelstra that IO crate should be unconditional but not `core2`.
